pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,324 @@
+	if d, ok := token.(json.Delim); ok && d == '[' {
+		// JSON array
+		return &jsonReader{decoder: decoder, token: token}, nil
+	}
+
+	// else reading the json object
+	return f.createJSONObjectReader()
+}
+
+// peekJSONToken peeks the first JSON token without advancing the decoder.
+func (*ftpFile) peekJSONToken(decoder *json.Decoder) (json.Token, error) {
+	newDecoder := *decoder
+
+	token, err := newDecoder.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+// createJSONObjectReader creates a JSON reader for a JSON object.
+func (f *ftpFile) createJSONObjectReader() (datasource.RowReader, error) {
+	name := f.Name()
+
+	// Close the current file handle
+	if err := f.Close(); err != nil {
+		return nil, err
+	}
+
+	buffer := make([]byte, 1024)
+
+	// Retrieve the file as *ftp.Response
+	res, err := f.conn.Retr(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the response
+	n, err := res.Read(buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer = buffer[:n]
+
+	reader := bytes.NewReader(buffer)
+
+	// Create a JSON decoder
+	decoder := json.NewDecoder(reader)
+
+	return &jsonReader{decoder: decoder}, nil
+}
+
+// createTextCSVReader creates a text reader for reading text files.
+func (f *ftpFile) createTextCSVReader() datasource.RowReader {
+	// Read the file content into a buffer
+	buffer := make([]byte, 1024)
+
+	n, err := f.response.Read(buffer)
+	if err != nil {
+		f.logger.Errorf("failed to read text file: %v", err)
+	}
+
+	buffer = buffer[:n]
+
+	reader := bytes.NewReader(buffer)
+
+	return &textReader{
+		scanner: bufio.NewScanner(reader),
+		logger:  f.logger,
+	}
+}
+
+// Next checks if there is another JSON object available.
+func (j jsonReader) Next() bool {
+	return j.decoder.More()
+}
+
+// Scan decodes the next JSON object into the provided structure.
+func (j jsonReader) Scan(i interface{}) error {
+	return j.decoder.Decode(&i)
+}
+
+// Next checks if there is another line available in the text file.
+func (f textReader) Next() bool {
+	return f.scanner.Scan()
+}
+
+// Scan scans the next line from the text file into the provided pointer to string.
+func (f textReader) Scan(i interface{}) error {
+	switch target := i.(type) {
+	case *string:
+		*target = f.scanner.Text()
+		return nil
+	default:
+		return errNotPointer
+	}
+}
+
+// Close closes the FTP file connection.
+func (f *ftpFile) Close() error {
+	// Close the FTP response.
+	err := f.response.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	// Retrieve the file content from FTP server and read into p
+	r, err := f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range b {
+		p[i] = b[i]
+	}
```

Comment:
This could panic from an out-of-bounds access to `p[i]`.

But then, this should probably just be `n := copy(p, b)` anyways.